File: Ass_5_10.py
import sys
import re

class SegmentTree:
    def __init__(self, arr):
        self.arr = arr
        self.n = len(arr)
        self.min_tree = [0] * (4 * self.n)
        self.max_tree = [0] * (4 * self.n)
        if self.n > 0:
            self.build(0, 0, self.n - 1)

    def build(self, node, start, end):
        if start == end:
            self.min_tree[node] = self.arr[start]
            self.max_tree[node] = self.arr[start]
            return

        mid = (start + end) // 2
        left_child = 2 * node + 1
        right_child = 2 * node + 2

        self.build(left_child, start, mid)
        self.build(right_child, mid + 1, end)

        self.min_tree[node] = min(self.min_tree[left_child], self.min_tree[right_child])
        self.max_tree[node] = max(self.max_tree[left_child], self.max_tree[right_child])

    def query_min(self, node, start, end, l, r):
        if r < start or end < l:
            return float('inf')
        if l <= start and end <= r:
            return self.min_tree[node]
            
        mid = (start + end) // 2
        left_min = self.query_min(2 * node + 1, start, mid, l, r)
        right_min = self.query_min(2 * node + 2, mid + 1, end, l, r)
        return min(left_min, right_min)

    def query_max(self, node, start, end, l, r):
        if r < start or end < l:
            return float('-inf')
        if l <= start and end <= r:
            return self.max_tree[node]
            
        mid = (start + end) // 2
        left_max = self.query_max(2 * node + 1, start, mid, l, r)
        right_max = self.query_max(2 * node + 2, mid + 1, end, l, r)
        return max(left_max, right_max)

    def update(self, node, start, end, idx, val):
        if start == end:
            self.arr[idx] = val
            self.min_tree[node] = val
            self.max_tree[node] = val
            return

        mid = (start + end) // 2
        left_child = 2 * node + 1
        right_child = 2 * node + 2

        if start <= idx <= mid:
            self.update(left_child, start, mid, idx, val)
        else:
            self.update(right_child, mid + 1, end, idx, val)

        self.min_tree[node] = min(self.min_tree[left_child], self.min_tree[right_child])
        self.max_tree[node] = max(self.max_tree[left_child], self.max_tree[right_child])

def solve():
    # Read all input
    input_data = open("input.txt").read()
    if not input_data.strip():
        return
        
    # Extract all words/numbers, ignoring punctuation like (, ) and ,
    tokens = re.findall(r'[a-zA-Z]+|\d+', input_data)
    
    if not tokens: return
    
    n = int(tokens[0])
    temperatures = [int(x) for x in tokens[1:n+1]]
    
    seg_tree = SegmentTree(temperatures)
    
    # Process queries
    idx = n + 1
    while idx < len(tokens):
        cmd = tokens[idx]
        if cmd == "RangeMin":
            l, r = int(tokens[idx+1]), int(tokens[idx+2])
            print(seg_tree.query_min(0, 0, n - 1, l, r))
            idx += 3
        elif cmd == "RangeMax":
            l, r = int(tokens[idx+1]), int(tokens[idx+2])
            print(seg_tree.query_max(0, 0, n - 1, l, r))
            idx += 3
        elif cmd == "Update":
            target_idx, val = int(tokens[idx+1]), int(tokens[idx+2])
            seg_tree.update(0, 0, n - 1, target_idx, val)
            idx += 3
        else:
            idx += 1 # Safety fallback

if __name__ == "__main__":
    solve()

File: Ass_5_10.java
import java.util.*;

public class Ass_5_10 {

    static class SegmentTree {
        int[] minTree;
        int[] maxTree;
        int[] arr;
        int n;

        public SegmentTree(int[] inputArr) {
            this.arr = inputArr;
            this.n = arr.length;
            // The maximum size of a segment tree array is 4 * N
            minTree = new int[4 * n];
            maxTree = new int[4 * n];
            build(0, 0, n - 1);
        }

        // 1. Build the Segment Tree
        private void build(int node, int start, int end) {
            if (start == end) {
                minTree[node] = arr[start];
                maxTree[node] = arr[start];
                return;
            }

            int mid = start + (end - start) / 2;
            int leftChild = 2 * node + 1;
            int rightChild = 2 * node + 2;

            build(leftChild, start, mid);
            build(rightChild, mid + 1, end);

            minTree[node] = Math.min(minTree[leftChild], minTree[rightChild]);
            maxTree[node] = Math.max(maxTree[leftChild], maxTree[rightChild]);
        }

        // 2. Perform Range Minimum Query
        public int queryMin(int node, int start, int end, int l, int r) {
            if (r < start || end < l) {
                return Integer.MAX_VALUE; // Return infinity so it doesn't affect min
            }
            if (l <= start && end <= r) {
                return minTree[node];
            }
            
            int mid = start + (end - start) / 2;
            int leftMin = queryMin(2 * node + 1, start, mid, l, r);
            int rightMin = queryMin(2 * node + 2, mid + 1, end, l, r);
            
            return Math.min(leftMin, rightMin);
        }

        // 3. Perform Range Maximum Query
        public int queryMax(int node, int start, int end, int l, int r) {
            if (r < start || end < l) {
                return Integer.MIN_VALUE; // Return -infinity so it doesn't affect max
            }
            if (l <= start && end <= r) {
                return maxTree[node];
            }
            
            int mid = start + (end - start) / 2;
            int leftMax = queryMax(2 * node + 1, start, mid, l, r);
            int rightMax = queryMax(2 * node + 2, mid + 1, end, l, r);
            
            return Math.max(leftMax, rightMax);
        }

        // 4. Update a sensor reading dynamically
        public void update(int node, int start, int end, int idx, int val) {
            if (start == end) {
                arr[idx] = val;
                minTree[node] = val;
                maxTree[node] = val;
                return;
            }

            int mid = start + (end - start) / 2;
            int leftChild = 2 * node + 1;
            int rightChild = 2 * node + 2;

            if (start <= idx && idx <= mid) {
                update(leftChild, start, mid, idx, val); 
            } else {
                update(rightChild, mid + 1, end, idx, val); 
            }

            minTree[node] = Math.min(minTree[leftChild], minTree[rightChild]);
            maxTree[node] = Math.max(maxTree[leftChild], maxTree[rightChild]);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        if (sc.hasNextInt()) {
            int n = sc.nextInt();
            int[] temperatures = new int[n];

            for (int i = 0; i < n; i++) {
                temperatures[i] = sc.nextInt();
            }

            SegmentTree segTree = new SegmentTree(temperatures);

            // Read queries until there is no more input
            while (sc.hasNext()) {
                // Read the query type string (e.g., "RangeMin(2,6)" or just "RangeMin" depending on formatting)
                // For simplicity, let's assume the input is spaced out like: RangeMin 2 6
                String rawInput = sc.next(); 
                
                // If your input literally has brackets like "RangeMin(2,6)", we clean it:
                if (rawInput.contains("(")) {
                    String clean = rawInput.replace("(", " ").replace(")", " ").replace(",", " ");
                    Scanner stringReader = new Scanner(clean);
                    String type = stringReader.next();
                    
                    if (type.equals("RangeMin")) {
                        System.out.println(segTree.queryMin(0, 0, n - 1, stringReader.nextInt(), stringReader.nextInt()));
                    } else if (type.equals("RangeMax")) {
                        System.out.println(segTree.queryMax(0, 0, n - 1, stringReader.nextInt(), stringReader.nextInt()));
                    } else if (type.equals("Update")) {
                        segTree.update(0, 0, n - 1, stringReader.nextInt(), stringReader.nextInt());
                    }
                    stringReader.close();
                } else {
                    // Standard space-separated input format
                    String type = rawInput;
                    if (type.equals("RangeMin")) {
                        System.out.println(segTree.queryMin(0, 0, n - 1, sc.nextInt(), sc.nextInt()));
                    } else if (type.equals("RangeMax")) {
                        System.out.println(segTree.queryMax(0, 0, n - 1, sc.nextInt(), sc.nextInt()));
                    } else if (type.equals("Update")) {
                        segTree.update(0, 0, n - 1, sc.nextInt(), sc.nextInt());
                    }
                }
            }
        }
        sc.close();
    }
}